README
NAME
    PlugAuth::Plugin::LDAP - LDAP back end for PlugAuth

VERSION
    version 0.06

SYNOPSIS
    In your PlugAuth.conf file:

     ---
     ldap :
       server : ldap://198.118.255.141:389
       dn : uid=%s, ou=people, dc=users, dc=example, dc=com
       authoritative : 1

    Note that %s in the dn will be replaced with the username when binding
    to the LDAP server.

DESCRIPTION
    Handle authentication only from LDAP server. Everything else is handled
    by PlugAuth::Plugin::FlatAuth (e.g. authorization, groups, etc).

METHODS
  PlugAuth::Plugin::LDAP->check_credentials( $user, $password )
    Given a user and password, check to see if the password is correct.

SEE ALSO
    PlugAuth, PlugAuth::Routes, PlugAuth::Plugin::FlatAuth

AUTHOR
    Graham Ollis <gollis@sesda3.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by NASA GSFC.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.09      2018-03-11 14:27:59 -0400
  - Deprecated. See https://github.com/clustericious/Clustericious/issues/46

0.08      2016-07-20 15:27:00 -0400
  - Mojo 7.x compatability

0.07      2015-11-29 16:08:01 -0500
  - Updated tests to use YAML instead of JSON in preeration
    for the latter being deprecated.
  - Use Test::Clustericious::Cluster instead of Test::Mojo
  - Moved repository to clustericious org on github
    (updated meta appropriately)

0.06      2014-01-22 17:45:36 America/New_York
  - Mojo 4.50 compat

0.05      2013-07-06 17:58:32 America/New_York
  - update metadata

0.04      2013-05-30 15:18:19 America/New_York
  - Fixes for MSWin32

0.03      2013-05-10 13:04:22 America/New_York
  - Documentation

0.02      2013-04-23 17:13:04 America/New_York
  - use MakeMaker instead of ModuleBuild
  - use [ACPS::RPM]

0.01      2012-11-29 15:17:49 America/New_York
  - initial version


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "FindBin" => 0,
    "Module::Build" => "0.3601",
    "Test::Mojo" => 0,
    "Test::More" => 0
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "LDAP back end for PlugAuth",
  "dist_author" => [
    "Graham Ollis <gollis\@sesda2.com>"
  ],
  "dist_name" => "PlugAuth-Plugin-LDAP",
  "dist_version" => "0.01",
  "license" => "restrictive",
  "module_name" => "PlugAuth::Plugin::LDAP",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Log::Log4perl" => 0,
    "Net::LDAP" => 0,
    "PlugAuth" => 0,
    "PlugAuth::Role::Auth" => 0,
    "PlugAuth::Role::Plugin" => 0,
    "Role::Tiny::With" => 0,
    "perl" => "v5.10.0"
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/PlugAuth/Plugin/LDAP.pm
package PlugAuth::Plugin::LDAP;

# ABSTRACT: (Deprecated) LDAP back end for PlugAuth
# VERSION

=head1 SYNOPSIS

In your PlugAuth.conf file:

 ---
 ldap :
   server : ldap://198.118.255.141:389
   dn : uid=%s, ou=people, dc=users, dc=example, dc=com
   authoritative : 1

Note that %s in the dn will be replaced with the username
when binding to the LDAP server.

=head1 DESCRIPTION

B<NOTE>: This module has been deprecated, and may be removed on or after 31 December 2018.
Please see L<https://github.com/clustericious/Clustericious/issues/46>.

Handle authentication only from LDAP server.
Everything else is handled by L<PlugAuth::Plugin::FlatAuth>
(e.g. authorization, groups, etc).

=cut

use strict;
use warnings;
use v5.10;
use Net::LDAP;
use Log::Log4perl qw/:easy/;
use Role::Tiny::With;

with 'PlugAuth::Role::Plugin';
with 'PlugAuth::Role::Auth';

=head1 METHODS

=head2 PlugAuth::Plugin::LDAP-E<gt>check_credentials( $user, $password )

Given a user and password, check to see if the password is correct.

=cut

sub check_credentials {
    my ($class, $user,$pw) = @_;
    $user = lc $user;

    my $ldap_config = $class->global_config->ldap(default => '');

    if (!$ldap_config or !$ldap_config->{authoritative}) {
        # Check files first.
        return 1 if $class->deligate_check_credentials($user, $pw);
    }
    return 0 unless $ldap_config;
    my $server = $ldap_config->{server} or LOGDIE "Missing ldap server";
    my $ldap = Net::LDAP->new($server, timeout => 5) or do {
        ERROR "Could not connect to ldap server $server: $@";
        return 0;
    };
    my $orig = $user;
    my $extra = $user =~ tr/a-zA-Z0-9@._-//dc;
    WARN "Invalid username '$orig', turned into $user" if $extra;
    my $dn = sprintf($ldap_config->{dn},$user);
    my $mesg = $ldap->bind($dn, password => $pw);
    $mesg->code or return 1;
    INFO "Ldap returned ".$mesg->code." : ".$mesg->error;
    return 0;
}

1;

=head1 SEE ALSO

L<PlugAuth>, L<PlugAuth::Routes>, L<PlugAuth::Plugin::FlatAuth>

=cut


Makefile.PL

use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "LDAP back end for PlugAuth",
  "AUTHOR" => "Graham Ollis <gollis\@sesda3.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "PlugAuth-Plugin-LDAP",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "PlugAuth::Plugin::LDAP",
  "PREREQ_PM" => {
    "Log::Log4perl" => 0,
    "Net::LDAP" => 0,
    "PlugAuth" => 0,
    "PlugAuth::Role::Auth" => 0,
    "PlugAuth::Role::Plugin" => 0,
    "Role::Tiny::With" => 0
  },
  "TEST_REQUIRES" => {
    "FindBin" => 0,
    "Test::Mojo" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.06",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "FindBin" => 0,
  "Log::Log4perl" => 0,
  "Net::LDAP" => 0,
  "PlugAuth" => 0,
  "PlugAuth::Role::Auth" => 0,
  "PlugAuth::Role::Plugin" => 0,
  "Role::Tiny::With" => 0,
  "Test::Mojo" => 0,
  "Test::More" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



